File: Experimental/2023/TrainingBot/src/main/cpp/commands/DriveDistance.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "commands/DriveDistance.h"
#include <iostream>

DriveDistance::DriveDistance(DriveBase* driveBase, units::meter_t distanceMeters, double speed) {
  // Use addRequirements() here to declare subsystem dependencies.
  AddRequirements(driveBase);

  m_driveBase = driveBase;
  m_distance = distanceMeters;
  m_speed = speed;
  m_targetDistance = 0_m;
}

// Called when the command is initially scheduled.
void DriveDistance::Initialize() {
  m_targetDistance = m_distance + m_driveBase->GetLeftDistance();
  m_driveBase->TankDrive(m_speed, m_speed);
  std::cerr << "Starting to drive forward "<< "\n";
}

// Called repeatedly when this Command is scheduled to run
void DriveDistance::Execute() {
  m_driveBase->TankDrive(m_speed, m_speed);
}

// Called once the command ends or is interrupted.
void DriveDistance::End(bool interrupted) {
  std::cerr << "Driving stopped. \n";
  m_driveBase->Stop();
}

// Returns true when the command should end.
bool DriveDistance::IsFinished() {
  if (m_driveBase->GetLeftDistance() >= m_targetDistance) {
    std::cerr <<"Finished driving."<< "\n";
    return true;
  }
  
  return false;
}

File: Experimental/2023/MattsToys/JavaUtilityLib/src/main/java/frc/robot/commands/DriveDistance.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import edu.wpi.first.wpilibj2.command.CommandBase;
import edu.wpi.first.wpilibj2.command.Subsystem;
import frc.robot.subsystems.DriveBaseInterface;

public class DriveDistance extends CommandBase {
  private final DriveBaseInterface m_drive;
  // Distance to drive in meters.
  private final double m_distance;
  private final double m_speed;
  private final boolean m_inReverse;
  private double m_targetDistance = 0;

  /**
   * Creates a new DriveDistance. This command will drive your your robot for a desired distance at
   * a desired speed.
   *
   * @param drive The drivetrain subsystem on which this command will run
   * @param speed The speed at which the robot will drive
   * @param meters The number of meters the robot will drive
   */
  public DriveDistance(DriveBaseInterface drive, double speed, double meters) {
    m_inReverse = (speed < 0 || meters < 0);
    final int sign = (m_inReverse ? -1 : +1);
    m_distance = Math.abs(meters) * sign;
    m_speed = Math.abs(speed) * sign;
    m_drive = drive;
    addRequirements((Subsystem) drive);
  }

  // Called when the command is initially scheduled.
  @Override
  public void initialize() {
    m_drive.tankDrive(0, 0);
    m_targetDistance = m_drive.getLeftEncoderPositionMeters() + m_distance;
  }

  // Called every time the scheduler runs while the command is scheduled.
  @Override
  public void execute() {
    m_drive.tankDrive(m_speed, m_speed);
  }

  // Called once the command ends or is interrupted.
  @Override
  public void end(boolean interrupted) {
    m_drive.stop();
  }

  // Returns true when the command should end.
  @Override
  public boolean isFinished() {
    final double currentDistance = m_drive.getLeftEncoderPositionMeters();
    return (m_inReverse && currentDistance <= m_targetDistance)
        || (!m_inReverse && currentDistance >= m_targetDistance);
  }
}
